sycl/source/backend/opencl.cpp

```diff
@@ -44,6 +46,66 @@ __SYCL_EXPORT queue make_queue(const context &Context,
   return detail::make_queue(NativeHandle, Context, nullptr, false,
                             ContextImpl->get_async_handler(), backend::opencl);
 }
+
+// Free functions to query OpenCL backend extensions
+// TODO: Extensions have been deprecated for aspects
+__SYCL_EXPORT bool has_extension(const sycl::platform &SyclPlatform,
+                                 const std::string &Extension) {
+  if (SyclPlatform.get_backend() != sycl::backend::opencl) {
+    throw sycl::exception(
+        errc::backend_mismatch,
+        "has_extension can only be used with an OpenCL backend");
+  }
+
+  auto PlatformImpl = getSyclObjImpl(SyclPlatform);
+  RT::PiPlatform PluginPlatform = PlatformImpl->getHandleRef();
+  plugin Plugin = PlatformImpl->getPlugin();
+
+  // Manual invocation of plugin API to avoid using deprecated
+  // info::platform::extensions call
+  size_t ResultSize = 0;
+  Plugin.call<PiApiKind::piPlatformGetInfo>(
+      PluginPlatform, PI_PLATFORM_INFO_EXTENSIONS, 0, nullptr, &ResultSize);
+  if (ResultSize == 0) {
+    return false;
+  }
+  std::unique_ptr<char[]> Result(new char[ResultSize]);
+  Plugin.call<PiApiKind::piPlatformGetInfo>(PluginPlatform,
+                                            PI_PLATFORM_INFO_EXTENSIONS,
+                                            ResultSize, Result.get(), nullptr);
+
+  std::string_view ExtensionsString(Result.get());
+  return ExtensionsString.find(Extension) != std::string::npos;
+}
+
+__SYCL_EXPORT bool has_extension(const sycl::device &SyclDevice,
+                                 const std::string &Extension) {
+  if (SyclDevice.get_backend() != sycl::backend::opencl) {
+    throw sycl::exception(
+        errc::backend_mismatch,
+        "has_extension can only be used with an OpenCL backend");
+  }
+
+  auto DeviceImpl = getSyclObjImpl(SyclDevice);
```

Comment:
FYI: [LLVM coding standard](https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable) says:

> Use `auto` if and only if it makes the code more readable or easier to maintain. Don’t “almost always” use `auto`...

I wouldn't say that type of `DeviceImpl` can be easily inferred just by looking at this line of code, but I don't have strong opinion here (and I'm not a codeowner here), so I will leave it up to others whether or not it is ok to use `auto` here.

Reply:
Changed to the full type of `std::shared_ptr<sycl::detail::device_impl>` (and platform equiv.) for clarity.